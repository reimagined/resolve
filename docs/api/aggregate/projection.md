---
id: projection
title: Projection
---

:::info TypeScript Support

An aggregate projection object has an associated TypeScript type:

- Type Name - `AggregateProjection`
- Containing Package - `@resolve-js/core`

:::

An aggregate projection is an object of the following structure:

```js
const projection = {
  // The *Init* function creates the initial aggregate state object.
  Init: () => initialState,
  // An event handler function is associated with an event type.
  // It receives the aggregate state and an incoming event
  // and returns the updated state.
  [EVENT_TYPE]: (state, event) => {
    ...
    return newState
  }
  [EVENT_TYPE2]: (state, event) => ...
  [EVENT_TYPE3]: (state, event) => ...
  ...
}
```

An event handler implementation receives the following arguments:

| Argument Name | Description                                                   |
| ------------- | ------------------------------------------------------------- |
| state         | The aggregate state that is an object of arbitrary structure. |
| event         | An [event](../event.md) object.                               |

An event handler should return a new state object.
