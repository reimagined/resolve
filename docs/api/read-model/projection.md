---
id: projection
title: Projection
---

:::info TypeScript Support

A read model projection has an associated TypeScript type:

- Type Name - `ReadModel`
- Package - `@resolve-js/core`

:::

A read model projection is an object of the following structure:

```js
const projection = {
  // The *Init* function initializes the read model's persistent store.
  Init: async (store) => { ... }
  // An event handler function is associated with an event type.
  // It receives the read model store and an incoming event
  // and updates the store based on the event's data.
  [EVENT_TYPE1]: async (store, event, context) -> { ... }
  [EVENT_TYPE2]: async (store, event, context) -> { ... }
}
```

An event handler implementation receives the following arguments:

| Argument Name | Description                                                                                              |
| ------------- | -------------------------------------------------------------------------------------------------------- |
| store         | Exposes [API](../read-model/store.md) used to communicate with the read model's persistent data storage. |
| event         | An object that contains the incoming event's data.                                                       |
| context       | An object that contains data and functions related to the current operation.                             |

## context

The `context` argument is an object with the following fields:

| Field Name | Description                                               |
| ---------- | --------------------------------------------------------- |
| encrypt    | The user-defined [encrypt](../../encryption.md) function. |
| decrypt    | The user-defined [decrypt](../../encryption.md) function. |

This object can also contain additional fields added by [middleware](../../middleware.md).
