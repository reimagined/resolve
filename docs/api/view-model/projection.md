---
id: projection
title: Projection
---

:::info TypeScript Support

A view model projection object has an associated TypeScript type:

- Type Name - `ViewModelProjection`
- Package - `@resolve-js/core`

:::

A view model projection is an object of the following structure:

```js
const projection = {
  // The *Init* function creates the view model's initial state object.
  Init: () => initialState,
  // An event handler function is associated with an event type.
  // It receives the view model's state and an incoming event
  // and returns the updated state.
  [EVENT_TYPE]: (state, event) => {
    ...
    return newState
  }
  [EVENT_TYPE2]: (state, event) => ...
  [EVENT_TYPE3]: (state, event) => ...
  ...
}
```

An event handler implementation receives the following arguments:

| Argument Name | Description                                                                                                                                    |
| ------------- | ---------------------------------------------------------------------------------------------------------------------------------------------- |
| state         | The view model's state that is an object of arbitrary structure.                                                                               |
| event         | An [event](../event.md) object.                                                                                                                |
| args          | Arguments attached to the request.                                                                                                             |
| context       | An object that contains functions and data related to the current operation (see the [Event Handler Context](#event-handler-context) section.) |

An event handler should return a new state object.

## Context

A view model event handler context is an object with the following fields:

| Field Name | Description                                               |
| ---------- | --------------------------------------------------------- |
| jwt        | The JSON Web Token attached to the request.               |
| encrypt    | The user-defined [encrypt](../../encryption.md) function. |
| decrypt    | The user-defined [decrypt](../../encryption.md) function. |
