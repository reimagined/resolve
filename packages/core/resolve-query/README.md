# **resolve-query**
[![npm version](https://badge.fury.io/js/resolve-query.svg)](https://badge.fury.io/js/resolve-query)

Provides an interface for creating and querying Read- and View-Models.

A **Read Model** is an application state built from the Event Log by Projection Functions and used to reply to Queries. It can be saved into a server-side Storage using an appropriate [Storage Adapter](../../adapters/storage-adapters). Queries to a Read Model are processed by Resolver Functions, which can filter, aggregate and/or paginate data before sending it to the client side. You can manage the Resolver behavior using client-defined arguments.  In this case, the server-side changes trigger the client-side updates.

A **View Model** is built on the client side and used to keep the UI up to the current application state. It is built and updated by Reducers. Aggregates whose events should participate in the View Model construction are defined by the `aggregateId` (in most cases, one aggregate corresponds to one view model, although, a wildcard `aggregateId` is possible).


```
import createQueryExecutor, { createReadModel, createViewModel } from 'resolve-query'

const queryExecutor = createQueryExecutor({
  eventStore,
  viewModels,
  readModels
})
```

## Usage
Pass the following arguments to the `createReadModel` factory function to create a **Read Model**:
* `eventStore` - A configured [eventStore](../resolve-es) instance.
* `projection` - An object with **Projection Functions** used to handle incoming Events and build the Read Model.
* `adapter` - A Read Model [adapter](../../adapters/readmodel-adapters) instance. A memory [adapter](../../adapters/readmodel-adapters/resolve-readmodel-lite), which supports a simple query & projection API is used by default.
* `resolvers` - An object with **Resolver Functions**. Each function should accept two arguments: the first argument is a reference to the target Read Model. The second argument is an object with fields used to extract data from a Read Model Storage.

A **Read Model** supports the following functions to query data:
* `read` - Performs a query within a Read Model via Resolvers. The first argument is a Resolver name; the second one contains an object with the Resolver function properties. Returns the Resolver execution result.
* `dispose` - Removes the Read Model. Invokes the Read Model Adapter's disposal mechanisms and stops listening to the Event Store bus. The default adapters' disposal operation disconnects with the storage and drops stored tables.


Pass the following arguments to the `createViewModel` factory function to create a **View Model**:
* `eventStore` - A configured [eventStore](../resolve-es) instance.
* `projection` - A map of [Redux-like Reducer functions](https://redux.js.org/docs/basics/Reducers.html). A function per each event type.
* `snapshotAdapter` - An adapter for loading and saving the intermediate Aggregate state. The argument is optional, snapshots are not used by default.

You can access Resolvers and query data from Read Models using a Facade, which supports the following functions:
* `read` - Performs a query from a View Model using Resolvers.
* `dispose` - Deletes the View Model (for internal use).

### Example
See the [with-saga](../../../examples/with-saga) example.

![Analytics](https://ga-beacon.appspot.com/UA-118635726-1/packages-resolve-query-readme?pixel)
