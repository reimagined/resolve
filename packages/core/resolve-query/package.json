{
  "name": "resolve-query",
  "version": "0.25.17",
  "description": "Provides an interface used to create and querry read models and view models in a reSolve application.",
  "engines": {
    "node": ">=10.9"
  },
  "main": "./lib/index.js",
  "module": "./es/index.js",
  "types": "./types/index.d.ts",
  "bugs": {
    "url": "https://github.com/reimagined/resolve/issues"
  },
  "scripts": {
    "build": "yarn workspace @internal/babel-compile start --name=resolve-query",
    "test": "jest --config=../../../jest.config.js --verbose",
    "testw": "jest --config=../../../jest.config.js --watchAll"
  },
  "repository": "https://github.com/reimagined/resolve.git",
  "author": "reimagined team",
  "license": "MIT",
  "keywords": [
    "cqrs",
    "eventsourcing",
    "es",
    "ddd"
  ],
  "homepage": "https://github.com/reimagined/resolve/tree/master/packages/core/resolve-query#readme",
  "dependencies": {
    "@babel/runtime": "7.9.6",
    "resolve-core": "0.25.17",
    "resolve-debug-levels": "0.25.17",
    "resolve-readmodel-base": "0.25.17"
  },
  "devDependencies": {
    "jest": "26.0.1"
  },
  "sourceType": "ts",
  "babelCompile": [
    {
      "sourceType": "ts",
      "moduleType": "es",
      "moduleTarget": "server",
      "inputDir": "./src",
      "outDir": "./es"
    },
    {
      "sourceType": "ts",
      "moduleType": "cjs",
      "moduleTarget": "server",
      "inputDir": "./src",
      "outDir": "./lib"
    }
  ]
}
