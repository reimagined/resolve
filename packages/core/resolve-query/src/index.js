import wrapReadModel from './wrap_read_model'
import wrapViewModel from './wrap_view_model'

const createQuery = ({
  readModelConnectors,
  snapshotAdapter,
  doUpdateRequest,
  readModels,
  viewModels,
  eventStore
}) => {
  const models = {}
  for (const readModel of readModels) {
    if (models[readModel.name] != null) {
      throw new Error(`Duplicate name for read/view model: ${readModel.name}`)
    }
    models[readModel.name] = wrapReadModel(
      readModel,
      readModelConnectors,
      doUpdateRequest
    )
  }

  for (const viewModel of viewModels) {
    if (models[viewModel.name] != null) {
      throw new Error(`Duplicate name for read/view model: ${viewModel.name}`)
    }
    models[viewModel.name] = wrapViewModel(
      viewModel,
      snapshotAdapter,
      eventStore
    )
  }

  const checkModelExists = modelName => {
    if (models[modelName] == null) {
      const error = new Error(`Read/view model ${modelName} does not exist`)
      error.code = 422
      throw error
    }
  }

  const parseOptions = options => {
    const optionsFlags = {
      modelOptions: 1 << 0,
      modelArgs: 1 << 1,
      resolverName: 1 << 2,
      resolverArgs: 1 << 3,
      aggregateIds: 1 << 4,
      aggregateArgs: 1 << 5
    }

    const optionsMap = []
    optionsMap[optionsFlags.modelOptions] = optionsMap[
      optionsFlags.modelOptions + optionsFlags.modelArgs
    ] = ['modelOptions', 'modelArgs']
    optionsMap[optionsFlags.resolverName] = optionsMap[
      optionsFlags.resolverName + optionsFlags.resolverArgs
    ] = ['resolverName', 'resolverArgs']
    optionsMap[optionsFlags.aggregateIds] = optionsMap[
      optionsFlags.aggregateIds + optionsFlags.aggregateArgs
    ] = ['aggregateIds', 'aggregateArgs']

    let flag = 0
    for (const key of Object.keys(options)) {
      flag += ~~optionsFlags[key]
    }

    if (optionsMap[flag] == null) {
      throw new Error('Wrong options for read invocation')
    }

    return [options[optionsMap[flag][0]], options[optionsMap[flag][1]]]
  }

  const read = async ({ modelName, jwtToken, ...options }) => {
    checkModelExists(modelName)
    const [modelOptions, modelArgs] = parseOptions(options)

    const result = await models[modelName].read(
      modelOptions,
      modelArgs,
      jwtToken
    )

    return result
  }

  const readAndSerialize = async ({ modelName, jwtToken, ...options }) => {
    checkModelExists(modelName)
    const [modelOptions, modelArgs] = parseOptions(options)

    const result = await models[modelName].readAndSerialize(
      modelOptions,
      modelArgs,
      jwtToken
    )

    return result
  }

  const updateByEvents = async (modelName, events) => {
    checkModelExists(modelName)
    if (!Array.isArray(events)) {
      throw new Error('Updating by events should supply events array')
    }

    await models[modelName].updateByEvents(events)
  }

  const drop = async modelName => {
    checkModelExists(modelName)

    await models[modelName].drop()
  }

  const dispose = async () => {
    for (const modelName of Object.keys(models)) {
      await models[modelName].dispose()
    }

    for (const connector of Object.values(readModelConnectors)) {
      await connector.disconnect()
    }
  }

  const api = {
    read,
    readAndSerialize,
    updateByEvents,
    drop,
    dispose
  }

  const executeQuery = (...args) => read(...args)
  Object.assign(executeQuery, api)

  return executeQuery
}

export default createQuery
