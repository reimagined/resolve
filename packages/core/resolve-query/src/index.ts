import wrapReadModel, {
  FULL_XA_CONNECTOR,
  FULL_REGULAR_CONNECTOR,
  EMPTY_CONNECTOR,
  INLINE_LEDGER_CONNECTOR,
  detectConnectorFeatures
} from './wrap-read-model'
import wrapViewModel from './wrap-view-model'

import { CreateQueryOptions } from './types'

const dispose = async (models: any): Promise<any> => {
  for (const modelName of Object.keys(models)) {
    await models[modelName].dispose()
  }
}

const interopApi = async (models: any, key: string, ...args: Array<any>) => {
  if (args.length !== 1 || Object(args[0]) !== args[0]) {
    throw new TypeError(
      `Invalid resolve-query method "${key}" arguments ${JSON.stringify(args)}`
    )
  }

  // eslint-disable-next-line prefer-const
  let { eventSubscriber, modelName, ...parameters } = args[0]

  if (eventSubscriber == null && modelName == null) {
    throw new Error(`Either "eventSubscriber" nor "modelName" is null`)
  } else if (modelName == null) {
    modelName = eventSubscriber
  }

  if (models[modelName] == null) {
    const error = new Error(
      `Read/view model "${modelName}" does not exist`
    ) as any
    error.code = 422
    throw error
  }

  const method = models[modelName][key]

  if (typeof method !== 'function') {
    throw new TypeError(
      `Model "${modelName}" does not implement method "${key}"`
    )
  }

  const result = await method(parameters)

  return result
}

const createQuery = (params: CreateQueryOptions): any => {
  const models: {
    [key: string]: any
  } = {}

  const { readModels, viewModels, ...imports } = params

  for (const readModel of readModels) {
    if (models[readModel.name] != null) {
      throw new Error(`Duplicate name for read model: "${readModel.name}"`)
    }
    models[readModel.name] = wrapReadModel({ readModel, ...imports })
  }

  for (const viewModel of viewModels) {
    if (models[viewModel.name] != null) {
      throw new Error(`Duplicate name for view model: "${viewModel.name}"`)
    }
    models[viewModel.name] = wrapViewModel({ viewModel, ...imports })
  }

  const read = interopApi.bind(null, models, 'read') as any
  const api = new Proxy(read, {
    get(_: any, key: string): any {
      if (key === 'bind' || key === 'apply' || key === 'call') {
        return read[key].bind(read)
      } else if (key === 'dispose') {
        return dispose.bind(null, models)
      } else {
        return interopApi.bind(null, models, key)
      }
    },
    set() {
      throw new TypeError(`Resolve-query API is immutable`)
    }
  } as any)

  return api
}

export {
  FULL_XA_CONNECTOR,
  FULL_REGULAR_CONNECTOR,
  EMPTY_CONNECTOR,
  INLINE_LEDGER_CONNECTOR,
  detectConnectorFeatures
}
export default createQuery
