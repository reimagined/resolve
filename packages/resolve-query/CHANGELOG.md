# Change Log

All notable changes to this project will be documented in this file.
See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.0.27"></a>
## 0.0.27 (2017-09-20)


### Features

* **resolve-query:** implement cross read-model graphql resolvers ([aed57ad](https://github.com/reimagined/resolve/commit/aed57ad))



<a name="0.0.2"></a>
## 0.0.2 (2017-05-30)




<a name="0.0.26"></a>
## 0.0.26 (2017-09-08)


### Features

* **resolve-query:** implement cross read-model graphql resolvers ([aed57ad](https://github.com/reimagined/resolve/commit/aed57ad))



<a name="0.0.2"></a>
## 0.0.2 (2017-05-30)




<a name="0.0.25"></a>
## 0.0.25 (2017-09-08)


### Features

* **resolve-query:** implement cross read-model graphql resolvers ([aed57ad](https://github.com/reimagined/resolve/commit/aed57ad))



<a name="0.0.2"></a>
## 0.0.2 (2017-05-30)
