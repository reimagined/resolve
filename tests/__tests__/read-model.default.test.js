import { createReadModel } from 'resolve-query'
import createMemoryReadModelAdapter from 'resolve-readmodel-memory'

import projection from '../read-models/default.projection'
import resolvers from '../read-models/default.resolvers'

describe('Read-model generic adapter API', () => {
  let buildTestReadModelReader, events

  it('Insert and non-parameterized resolver invocation', async () => {
    events.push({ type: 'INSERT_TEST', timestamp: 100, payload: 'test' })
    const reader = buildTestReadModelReader()

    const result = await reader('NON_PARAMETERIZED_RESOLVER_TEST', {})
    expect(result).toMatchSnapshot()
  })

  it('Update and non-parameterized resolver invocation', async () => {
    events.push({ type: 'INSERT_TEST', timestamp: 100, payload: 'test' })
    events.push({ type: 'UPDATE_TEST', timestamp: 101, payload: 'test' })
    const reader = buildTestReadModelReader()

    const result = await reader('NON_PARAMETERIZED_RESOLVER_TEST', {})
    expect(result).toMatchSnapshot()
  })

  it('Upsert and non-parameterized resolver invocation', async () => {
    events.push({ type: 'INSERT_TEST', timestamp: 100, payload: 'test' })
    events.push({ type: 'UPSERT_TEST', timestamp: 101, payload: 'test' })
    const reader = buildTestReadModelReader()

    const result = await reader('NON_PARAMETERIZED_RESOLVER_TEST', {})
    expect(result).toMatchSnapshot()
  })

  it('Delete and non-parameterized resolver invocation', async () => {
    events.push({ type: 'INSERT_TEST', timestamp: 100, payload: 'test' })
    events.push({ type: 'DELETE_TEST', timestamp: 101, payload: 'test' })
    const reader = buildTestReadModelReader()

    const result = await reader('NON_PARAMETERIZED_RESOLVER_TEST', {})
    expect(result).toMatchSnapshot()
  })

  it('Update and parameterized resolver invocation', async () => {
    events.push({ type: 'INSERT_TEST', timestamp: 100, payload: 'test' })
    events.push({ type: 'UPDATE_TEST', timestamp: 101, payload: 'test' })
    const reader = buildTestReadModelReader()

    const result = await reader('PARAMETRIZED_RESOLVER_TEST', {
      firstFieldCondition: 10,
      secondFieldCondition: 2,
      pageNumber: 2,
      pageLength: 5
    })
    expect(result).toMatchSnapshot()
  })

  beforeEach(async () => {
    events = []
    const eventStore = {
      async subscribeByEventType(eventTypes, handler, { startTime = 0 } = {}) {
        for (let event of events) {
          if (
            event &&
            eventTypes.indexOf(event.type) > -1 &&
            event.timestamp >= startTime
          ) {
            handler(event)
            await Promise.resolve()
          }
        }

        return () => null
      }
    }

    buildTestReadModelReader = () => {
      const readModel = createReadModel({
        adapter: createMemoryReadModelAdapter(),
        projection,
        eventStore,
        resolvers
      })

      return async (resolverName, resolverArgs) => {
        let result = await readModel.read({ resolverName, resolverArgs })
        const lastError = await readModel.getLastError()

        if (lastError != null) {
          throw lastError
        }

        return result
      }
    }
  })

  afterEach(async () => {
    buildTestReadModelReader = null
    events = null
  })
})
