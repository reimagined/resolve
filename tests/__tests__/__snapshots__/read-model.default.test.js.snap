// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Read-model generic adapter API Delete and non-parameterized resolver invocation 1`] = `
Object {
  "countResult": 3,
  "findOneResult": Object {
    "firstJsonName": Object {
      "a": 1,
      "b": 2,
      "e": 10,
    },
  },
  "findResult": Array [],
}
`;

exports[`Read-model generic adapter API Insert and non-parameterized resolver invocation 1`] = `
Object {
  "countResult": 4,
  "findOneResult": Object {
    "firstJsonName": Object {
      "a": 1,
      "b": 2,
      "e": 10,
    },
  },
  "findResult": Array [],
}
`;

exports[`Read-model generic adapter API Update and non-parameterized resolver invocation 1`] = `
Object {
  "countResult": 4,
  "findOneResult": Object {
    "firstJsonName": Object {
      "a": 1,
      "b": 2,
      "e": 10,
    },
  },
  "findResult": Array [],
}
`;

exports[`Read-model generic adapter API Update and parameterized resolver invocation 1`] = `
Array [
  Object {
    "firstJsonName": Object {
      "c": 3,
      "e": 25,
      "f": "inner-field",
    },
  },
  Object {
    "firstJsonName": null,
  },
]
`;

exports[`Read-model generic adapter API Upsert and non-parameterized resolver invocation 1`] = `
Object {
  "countResult": 5,
  "findOneResult": Object {
    "firstJsonName": Object {
      "a": 1,
      "b": 2,
      "e": 10,
    },
  },
  "findResult": Array [],
}
`;
